### ISSUE TITLE
Add process user name

### ISSUE THREAD

**Issue by owulff** [elastic/topbeat#36]
Add the process user name to the documents of type "proc". If a command like "java" is running several times with different local users you can differentiate that in the dashboard.

See discussion here:
https://discuss.elastic.co/t/add-process-user-name/29463/2?u=owulff

**Comment by ruflin** [comment]
I think this information could be found in https://godoc.org/github.com/cloudfoundry/gosigar#ProcExe

I removed the 1.0.0-beta4 label as I think this not critical for the release.

**Comment by owulff** [comment]
Honestly, this information is key for us because otherwise the process level view doesn't provide the information which Java process consumes more resources than others. Several applications run on the same box but with different users. Without the process user information, topbeat is not helpful to implement monitoring.

**Comment by urso** [comment]
I agree an username in the doc can be very useful, but relying on user name only to differentiate between multiple service is also limited in functionality.
How do you start your java processes. Have you considered e.g. launch4j? On unix systems you can rename an applications name via exec:

    $ exec -a myname java ...

The '-a' overwrite arg[0], which is the process its name. For example:

    # in first shell:
    $ bash  # start new bash to run exec from within (exec replaces current processes)
    bash-<version> $ exec -a test bash # replace current bash with bash named test

    # in second shell:
    $ ps aux | grep test | grep -v grep
    urso            10444   0.0  0.0  2446696   1396 s000  S+   11:28PM   0:00.01 test

Starting java processes via 'exec -a' gives you readable names in top, atop, htop and ps.

**Comment by ruflin** [comment]
@owulff I added the GA label. Lets see if we get it in for GA.

**Comment by monicasarbu** [comment]
I think this is easy enough and it can go in the next release, beta4.

**Comment by urso** [comment]
On linux (source code of proc: http://procps.sourceforge.net/download.html):
    1. read/parse /proc/[pid]/status
    2. "uid/gid" line contains: Real, effective, saved set, and filesystem UIDs separated by whitespace
    3. use getpwuid (try to cache results?) to get user name for uid

On OS X (see [ps command source code](http://www.opensource.apple.com/source/adv_cmds/adv_cmds-149/ps/ps.c)): 
  1. find uid from pid using
  either [sysctl](http://stackoverflow.com/questions/6457682/how-to-programatically-get-uid-from-pid-in-osx-using-c)
  or try libproc (http://opensource.apple.com/source/xnu/xnu-1699.24.23/bsd/kern/proc_info.c). See header include/libproc.h on XCode (On my machine /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include)
  2. use getpwuid (try to cache results?) to get user name for uid

Windows:
  maybe that's an options: http://stackoverflow.com/questions/2686096/c-get-username-from-process

**Comment by monicasarbu** [comment]
Adding username information is easy on linux, but relatively complex on OSX and Windows. Sadly, it will not make it in the beta4 release. 

@owulff Do you think having the full command line for each process would solve your issue? I am thinking this would be useful for more users.

**Comment by owulff** [comment]
Linux would be perfectly fine for now. Windows maybe later, but no need for OSX support.

The command line doesn't contain the user name. The user name and the binary name are the key information we look at to implement alerting and such.

**Comment by ei-grad** [comment]
@monicasarbu full command line looks too heavy, and if we talk about Java... maybe adding a tag for regexp on full command line of process could be not complicated to implement?

**Comment by monicasarbu** [comment]
@ei-grad Your proposal sounds like conditional filtering, a feature that we are planning to start implementing after we release 1.0.0-GA.

**Comment by owulff** [comment]
Is there any progress on this requests. Initially planned for beta4, then GA and now unknown.

As of now, topbeat helps us only to get an indication about the status of the server itself. We must know which java process is consuming too much CPU.

We have run a test with exec -a in our LAB environment but it didn't work.

**Comment by monicasarbu** [comment]
We didn't start working on this. We will try implementing only for Linux in the next release, but I cannot  promise anything now.

**Comment by LucaWintergerst** [comment]
+1 :)

**Comment by cmfatih** [comment]
+1

**Comment by biswajit86** [comment]
+1.
Do you have an idea when this feature would be in..  i am fine testing it out in a beta version as well

**Comment by dr-toboggan** [comment]
+1 We are interested in this feature on Windows and willing to contribute code to implement it.  I think we can get the process username by using sigar to query the environment variables of the process to get the "USERNAME" and "USERDOMAIN" variables.  Please let me know if you have any guidance/suggestions to keep in mind with the implementation or to maintain consistency with any work already done on the Linux side.

**Comment by ruflin** [comment]
@dr-toboggan It would be really great if you could help here contributing. As the implementation wasn't done yet, it is best you go with what you would expect and we discuss the details then on a PR.

**Comment by eonarheim** [comment]
I've committed a potential fix for user name information on my fork of gosigar, I'm just going to confirm it works when rolled into topbeat on windows.

I'm not sure which repo should be tracking these issues "beats" or "topbeat" let me know :)
